cluster-autoscaler/estimator/binpacking_estimator.go

```diff
@@ -105,47 +110,66 @@ func (e *BinpackingNodeEstimator) Estimate(
-			// on a new node either. There is no point adding more nodes to snapshot in such case, especially because of
-			// performance cost each extra node adds to future FitsAnyNodeMatching calls.
-			if lastNodeName != "" && !newNodesWithPods[lastNodeName] {
-				continue
+		for ; index < len(podsEquivalenceGroup.Pods); index++ {
+			pod := podsEquivalenceGroup.Pods[index]
+			found := false
+
+			// Check schedulability on only newly created node
+			nodeName, err := e.predicateChecker.FitsAnyNodeMatching(e.clusterSnapshot, pod, func(nodeInfo *schedulerframework.NodeInfo) bool {
+				return nodeInfo.Node().Name == lastNodeName
+			})
+			if err == nil {
+				found = true
+				if err := e.clusterSnapshot.AddPod(pod, nodeName); err != nil {
+					klog.Errorf("Error adding pod %v.%v to node %v in ClusterSnapshot; %v", pod.Namespace, pod.Name, nodeName, err)
+					return 0, nil
+				}
+				scheduledPods = append(scheduledPods, pod)
+				newNodesWithPods[nodeName] = true
 			}
 
-			// Stop binpacking if we reach the limit of nodes we can add.
-			// We return the result of the binpacking that we already performed.
-			//
-			// The thresholdBasedEstimationLimiter implementation assumes that for
-			// each call that returns true, one node gets added. Therefore this
-			// must be the last check right before really adding a node.
-			if !e.limiter.PermissionToAddNode() {
-				break
+			if !found {
+				// If the last node we've added is empty and the pod couldn't schedule on it, it wouldn't be able to schedule
+				// on a new node either. There is no point adding more nodes to snapshot in such case, especially because of
+				// performance cost each extra node adds to future FitsAnyNodeMatching calls.
+				if lastNodeName != "" && !newNodesWithPods[lastNodeName] {
+					continue
+				}
+
+				// Stop binpacking if we reach the limit of nodes we can add.
+				// We return the result of the binpacking that we already performed.
+				//
+				// The thresholdBasedEstimationLimiter implementation assumes that for
+				// each call that returns true, one node gets added. Therefore this
+				// must be the last check right before really adding a node.
+				if !e.limiter.PermissionToAddNode() {
+					break
+				}
+
+				// Add new node
+				newNodeName, err := e.addNewNodeToSnapshot(nodeTemplate, newNodeNameIndex)
+				if err != nil {
+					klog.Errorf("Error while adding new node for template to ClusterSnapshot; %v", err)
+					return 0, nil
+				}
+				newNodeNameIndex++
+				newNodeNames[newNodeName] = true
+				lastNodeName = newNodeName
+
+				// And try to schedule pod to it.
+				// Note that this may still fail (ex. if topology spreading with zonal topologyKey is used);
+				// in this case we can't help the pending pod. We keep the node in clusterSnapshot to avoid
+				// adding and removing node to snapshot for each such pod.
+				if err := e.predicateChecker.CheckPredicates(e.clusterSnapshot, pod, newNodeName); err != nil {
+					continue
```

Comment:
Shouldn't we skip the entire group in this case? If one pod failed, what's the point of trying another identical pod?

Reply:
Done.